### PR TITLE
Rewrite resolver docs

### LINKED ISSUE #5237: Universal resolution needs documentation
e.g in `docs/resolution.md#multi-platform-resolution` or in a new section after it.

### LINKED ISSUE #5238: Resolution concept documentation needs introduction
What is resolution and why do we do it? at the top of `docs/resolution.md`.

### LINKED ISSUE #5603: Write documentation for the resolver
We should write a better explanation for the how the resolver and the lockfile work.

### PR DESCRIPTION
**Pull request by konstin**
This PR rewrites the resolver concept and adds a resolver internals page targeted at power users.

The new resolution concept documentation has three parts:
* An introduction for people how never heard of "resolution" before, and a motivating example what it does. I've also shoved the part about equally valid resolutions in there.
* Features you commonly use: Non-universal vs. universal resolution, lowest resolution amd pre-releases.
* Expert features, we don't advertise them, you'll only need them in complex cases when you already know and i kept them to the reference points you need in this case: Constraints, overrides and exclude-newer.

I intentionally didn't lay out any detail of the resolution itself, the idea is that users get a vague sense of "uv has to select fitting versions", but then they learn the options they have to use and some common failure points without ever coming near SAT or even graphs.

The resolution internals reference page is targeted at power users who need to understand what is going on behind the scenes. It assumes ample prior knowledge and exists to explain the uv-specific behaviors for someone who already understands dependency resolution conceptually and has interacted with their dependency tree before. I had a section on the lockfile but removed it because it found the lockfile to be too self-documenting.

I haven't touched the readme.

Closes #5603
Closes #5238
Closes #5237

### CODE REVIEW

File: BENCHMARKS.md [hunk 1]
```diff
@@ -9,11 +9,11 @@ important caveats:
 - Benchmark performance may vary dramatically depending on the set of packages being installed. For
   example, a resolution that requires building a single intensive source distribution may appear
   very similar across tools, since the bottleneck is tool-agnostic.
-- Unlike Poetry, both uv and pip-tools do _not_ generate multi-platform lockfiles. As such, Poetry
-  is (by design) doing significantly more work than other tools in the resolution benchmarks. Poetry
-  is included for completeness, as many projects may not _need_ a multi-platform lockfile. However,
-  it's critical to understand that benchmarking uv's resolution time against Poetry is an unfair
-  comparison. (Benchmarking installation, however, _is_ a fair comparison.)
+- Unlike Poetry, both uv and pip-tools do _not_ generate platform-independent lockfiles. As such,
```
Comment (zanieb) [inline BENCHMARKS.md:12]: We will need to update this separately.

**Review by zanieb — APPROVED** [review]
Thanks! I'll post edits separately.

File: docs/concepts/resolution.md [hunk 2]
```diff
@@ -1,20 +1,88 @@
...
+
+We can't install both `foo` 2.0.0 and `bar` 2.0.0 because they conflict on the version of `lib`, so
+the resolver will pick either `foo` 1.0.0 or `bar` 1.0.0. Both are valid solutions, at the resolvers
+choice.
+
+## Platform-specific and universal resolution
+
+uv supports two modes of resolution: Platform-specific and universal (platform-independent).
+
+Like `pip` and `pip-tools`, `uv pip compile` produces a resolution that's only known to be
```
Comment (konstin) [inline docs/concepts/resolution.md:25]: There is overlap with the pip page here but it's hard not to have that

File: docs/concepts/resolution.md [hunk 3]
```diff
@@ -1,20 +1,88 @@
...
+to use, not a lower bound.
+
+!!! note
+
+    Python's environment markers expose far more information about the current machine
+    than can be expressed by a simple `--python-platform` argument. For example, the `platform_version` marker
+    on macOS includes the time at which the kernel was built, which can (in theory) be encoded in
+    package requirements. uv's resolver makes a best-effort attempt to generate a resolution that is
+    compatible with any machine running on the target `--python-platform`, which should be sufficient for
+    most use cases, but may lose fidelity for complex package and platform combinations.
```
[the remark this answers is not in the document]
Comment (BurntSushi) [reply docs/concepts/resolution.md:56]: This is somewhat addressed in the "expert" docs below. But it might be worth a brief mention here.

**Comment by zanieb** [comment]
I'll take @BurntSushi's comments into account.

File: docs/concepts/resolution.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (konstin) [reply docs/concepts/resolution.md:56]: I think this something we should communicate in the error message (specifically, add a note in the build error when we're on a different platform), it's not something you need to think upfront in the concept docs

**Pull request merged**